### PR TITLE
added reflection closes #93

### LINKED ISSUE #93: Reflection on the product and process from a software engineering perspective

### PR DESCRIPTION
**Pull request by GijsWeterings**

### CODE REVIEW

File: Documentation/General/finalReport/reflection.tex [hunk 1]
```diff
@@ -1 +1,107 @@
...
+The API in turn talks to a persistence server in the form of a PostgreSQL database.
+
+\section{Design Patterns}
+A lot of the design patterns in our application come from the libraries and frameworks we use. However, we did design our backend system very methodically, using the model-view-controller (MVC) pattern. The API classes define the view of the application, in the form of JSON responses, which is what our frontend uses to get it's data. The "backend" package is what defines the controller of the system, and the DAO's (Data Access Object) define our models. They talk to the postgreSQL database for retrieval and persistence.
+
+In the frontend we use a slightly different structure, most commonly referred to MVVM (Model, View, ViewModel). The models here are the API calls, controllers and services are the viewmodels, and the HTML templates are the views.
+
+A few of the design patterns we used:
+\begin{enumerate}
+\item facade
```
Comment (jwgmeligmeyling) [inline Documentation/General/finalReport/reflection.tex]: Maybe write all design patterns with a capital for consistency.

File: Documentation/General/finalReport/reflection.tex [hunk 2]
```diff
@@ -1 +1,107 @@
...
+\item singleton
+\item factories
+\item Dependency injection
+\item MVC
+\item MVVM
+\item Resource pool
+\end{enumerate}
+
+\section{Code readability / maintainability}
+One of our goals in this project was creating a readable and maintainable codebase.
```
Comment (jwgmeligmeyling) [inline Documentation/General/finalReport/reflection.tex]: Imo Lombok is mostly a development efficiency improver, and the code readability benefit is quite minor. While our system is really developed with maintainability and readability in mind, I think we should focus on a few other properties in this paragraph:
- We use small method bodies with descriptive names
- We use lamba's and streams, which are way more readable than loops, nested ifs, unnecessary copies and above all, anonymous classes.
- We use QueryDSL to write queries using a builder pattern, which makes queries more readable and adds the benefit of type checking for our queries.
- We use Google Guice which really allows us to rapid prototype classes and use composition at its best and move dependencies around in refactoring.
- We use various contracts and classes for the interactions between the database layer, backend layer and API layer, which thus can be seen as a layered model, which add the benefit that these layers can be individually maintained while the other layers will keep working.

File: Documentation/General/finalReport/reflection.tex [hunk 3]
```diff
@@ -1 +1,107 @@
...
+
+To replace our initial plans of a Neural Network, we're using rtrees, which are spatial indices for objects.
+To get support for that in our application, we're using Hibernate Spatial.
+
+For development, we're using two more major dependencies.
+To simulate a postgres database during testing we're using H2 (with Hatbox for spatial query support).
+Besides that, we're using Project Lombok for automatic generation of getters, setters and hashcode and equals methods.
+
+\subsection{Persistence}
+In the persistence layer, we're using postgreSQL as the database, with postgre Spatial as a plugin for spatial support.
```
Comment (jwgmeligmeyling) [inline Documentation/General/finalReport/reflection.tex]: Maybe it ends somewhat abruptly here, I will think about this...

File: Documentation/General/finalReport/reflection.tex [hunk 4]
```diff
@@ -1 +1,107 @@
...
+This is quite a controversial project, as it intercepts the code on compile time, and then injects getters, setters, and hashcode/equals methods.
+These are not in the source files at all, and with the proper IDE plugin we still have autocomplete.
+The result of all this is a source file without a lot of boilerplate methods.
+This is way easier to read and understand.
+This speeds up both development and learning the codebase.
+
+Furthermore, we've set on creating the API in a way such that a small change in backend code doesn't have to affect frontend. Adding a field to an object for example, is something that may introduce a bug in a clientside application. 
+
+\section{SCRUM}
+We worked with SCRUM for the first time during this project.
```
Comment (eanker) [inline Documentation/General/finalReport/reflection.tex]: We have used it with SEM.
You could say big project, or something like that.

File: Documentation/General/finalReport/reflection.tex (hunk 3 above)
Comment (GijsWeterings) [inline Documentation/General/finalReport/reflection.tex]: wel did you? >:-|

No I agree... Maybe remove this subsection altogether?

**Comment by jwgmeligmeyling** [comment]
Good! :+1:

**Comment by TimvdLippe** [comment]
Did quite some rewording, but overall structure was good!

**Comment by jwgmeligmeyling** [comment]
Merge the shit out of this pr

**Pull request merged**